packages/web-components/fast-ssr/src/template-parser/template-parser.ts

```diff
@@ -0,0 +1,351 @@
+
+    if ("childNodes" in node) {
+        const { childNodes } = node;
+        for (const child of childNodes) {
+            traverse(child, visitor);
+        }
+    }
+
+    if (visitor.leave) {
+        visitor.leave(node);
+    }
+}
+
+/**
+ * Test if a node is a comment node.
+ * @param node - the node to test
+ */
+function isCommentNode(node: DefaultTreeNode): node is DefaultTreeCommentNode {
+    return node.nodeName === "#comment";
+}
+
+/**
+ * Test if a node is a text node.
+ * @param node - the node to test
+ */
+function isTextNode(node: DefaultTreeNode): node is DefaultTreeTextNode {
+    return node.nodeName === "#text";
+}
+
+/**
+ * Test if a node is an element node
+ * @param node - the node to test
+ */
+function isElementNode(node: DefaultTreeNode): node is DefaultTreeElement {
+    return (node as DefaultTreeElement).tagName !== undefined;
+}
+
+/**
+ * Determines which type of attribute binding an attribute is
+ * @param attr - The attribute to inspect
+ */
+function getAttributeType(attr: Attribute): AttributeType {
+    const result = attributeTypeRegExp.exec(attr.name);
+
+    if (result === null) {
+        throw new Error("Failure to determine attribute binding type");
+    }
+
+    const prefix = result[1];
+
+    return prefix === ":"
+        ? AttributeType.idl
+        : prefix === "?"
+        ? AttributeType.booleanContent
+        : prefix === "@"
+        ? AttributeType.event
+        : AttributeType.content;
+}
+
+/**
+ * Tests if a node is an interpolated FAST marker
+ * @param node - the node to test
+ */
+function isInterpolationMarkerNode(
+    node: DefaultTreeNode
+): node is Required<DefaultTreeTextNode> {
+    return (
+        isTextNode(node) &&
+        node.sourceCodeLocation !== undefined &&
+        isInterpolationMarker(node)
+    );
+}
+
+/**
+ * Tests if a node is a FAST comment marker
+ * @param node - the node to test
+ */
+function isCommentMarkerNode(
+    node: DefaultTreeNode
+): node is Required<DefaultTreeCommentNode> {
+    return (
+        isCommentNode(node) &&
+        node.sourceCodeLocation !== undefined &&
+        isMarkerComment(node)
+    );
+}
+
+/**
+ * Parses a template into a set of operation instructions
+ * @param template - The template to parse
+ */
+export function parseTemplateToOpCodes(template: ViewTemplate): Op[] {
+    const cached: Op[] | undefined = opCache.get(template);
+    if (cached !== undefined) {
+        return cached;
+    }
+
+    const { html } = template;
+
+    if (typeof html !== "string") {
+        throw new Error(
+            "@microsoft/fast-ssr does not support rendering a ViewTemplate with an HTMLTemplateElement html source."
+        );
+    }
+
+    /**
+     * Typescript thinks that `html` is a string | HTMLTemplateElement inside the functions defined
+     * below, so store in a new var that is just a string type
+     */
+    const templateString = html;
+    const ast = parseFragment(html, { sourceCodeLocationInfo: true });
```

Comment:
ast = a string type? Can we spell it out:

```suggestion
    const aStringType = parseFragment(html, { sourceCodeLocationInfo: true });
```

Reply:
abstract syntax tree ;) I can find a more descriptive name.